### PR TITLE
Allow using filter when materializing BigQuery views

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Allow using filter when materializing BigQuery views to avoid generating huge temporary tables. 

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# BigQuery
* TBD. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplit.java [hunk 1]
```diff
@@ -60,26 +62,27 @@ public BigQuerySplit(
...
+        this.tableName = requireNonNull(tableName, "tableName is null");
         this.schemaString = requireNonNull(schemaString, "schemaString cannot be null");
         this.columns = ImmutableList.copyOf(requireNonNull(columns, "columns cannot be null"));
         this.emptyRowsToGenerate = emptyRowsToGenerate;
         this.filter = requireNonNull(filter, "filter is null");
         this.dataSize = requireNonNull(dataSize, "dataSize is null");
     }
 
-    static BigQuerySplit forStream(String streamName, String schemaString, List<BigQueryColumnHandle> columns, OptionalInt dataSize)
+    static BigQuerySplit forStream(String streamName, String tableName, String schemaString, List<BigQueryColumnHandle> columns, OptionalInt dataSize)
```
Comment (wendigo) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplit.java:73]: Why the tableName is before schemaString? reorder. Can we use SchemaTableName here instead?

Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplit.java]: schemaString is string of table definition, not schema name. 

This value holds 3 level identifier (e.g. `projects/development-226103/datasets/test/tables/_pbc_95551a80eb2441d7a343153d7efc5887`), so SchemaTableName looks not suitable. Changed to iamResourceName.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQuerySplitManager.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    private final BigQuerySqlExecutor bigQueryExecutor;
+
+    public TestBigQuerySplitManager()
+    {
+        BigQueryConnectorFactory connectorFactory = new BigQueryConnectorFactory();
+        connector = connectorFactory.create("bigquery", ImmutableMap.of("bigquery.views-enabled", "true"), new TestingConnectorContext());
+        bigQueryExecutor = new BigQuerySqlExecutor();
+    }
+
+    @Test
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQuerySplitManager.java:58]: This test is very specific to BigQuerySplitManager. Separated class is better in my opinion.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQuerySplitManager.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        ConnectorMetadata metadata = connector.getMetadata(session, transaction);
+
+        String materializedView = "test_materialized_view" + randomNameSuffix();
+        onBigQuery("CREATE MATERIALIZED VIEW test." + materializedView + " AS SELECT count(1) AS cnt FROM tpch.region");
+        try {
+            BigQueryTableHandle table = (BigQueryTableHandle) metadata.getTableHandle(session, new SchemaTableName("test", materializedView), Optional.empty(), Optional.empty());
+
+            BigQuerySplit split = getBigQuerySplit(transaction, session, table);
+            assertThat(split.getMode()).isEqualTo(STORAGE);
+            assertThat(split.getIamResourceName()).contains("_pbc_");
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQuerySplitManager.java:75]: https://github.com/trinodb/trino/blob/ca3503bb172abdb658170c8675f240e496502983/plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/ViewMaterializationCache.java#L73

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQuerySplitManager.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+            // Reuse the same temporary table regardless of filter by default
+            BigQueryColumnHandle column = new BigQueryColumnHandle("cnt", BIGINT, INT64, true, REQUIRED, ImmutableList.of(), null, false);
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQuerySplitManager.java:78]: This variable is used from single place. Moving to constant doesn't bring any benefit in my opinion.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQuerySplitManager.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (wendigo) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQuerySplitManager.java:75]: Let's extract _pbc_ to a class const

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQuerySplitManager.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            throws Exception
+    {
+        ConnectorSession session = TestingConnectorSession.builder()
+                .setPropertyMetadata(new BigQuerySessionProperties(new BigQueryConfig()).getSessionProperties())
+                .build();
+        ConnectorTransactionHandle transaction = connector.beginTransaction(READ_UNCOMMITTED, false, true);
+        ConnectorMetadata metadata = connector.getMetadata(session, transaction);
+
+        String materializedView = "test_materialized_view" + randomNameSuffix();
+        onBigQuery("CREATE MATERIALIZED VIEW test." + materializedView + " AS SELECT count(1) AS cnt FROM tpch.region");
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQuerySplitManager.java:69]: We can't use it at this time. I would stay with the current code in this PR.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQuerySplitManager.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        try {
+            BigQueryTableHandle table = (BigQueryTableHandle) metadata.getTableHandle(session, new SchemaTableName("test", materializedView), Optional.empty(), Optional.empty());
+
+            BigQuerySplit split = getBigQuerySplit(transaction, session, table);
+            assertThat(split.getMode()).isEqualTo(STORAGE);
+            assertThat(split.getIamResourceName()).contains("_pbc_");
+
+            // Reuse the same temporary table regardless of filter by default
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/TestBigQuerySplitManager.java]: Changed to "// Ignore constraints when creating temporary tables by default (view_materialization_with_filter is false)"

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplit.java [hunk 7]
```diff
@@ -94,6 +97,12 @@ public String getStreamName()
         return streamName;
     }
 
+    @JsonProperty
```
[the remark this answers is not in the document]
Comment (wendigo) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplit.java:100]: Can we instead inspect ViewMaterializationCache ? As far as I understand when `bigqueryview-materialization-with-filter` is enabled and we pass a filter to the materialized view, we should materialize view multiple times if filter is different, right?

[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplit.java]: Inspecting ViewMaterializationCache will require some changes around injector in the connector. 

I refactored to expose ReadSession in BigQuerySplitManager instead.

[the remark this answers is not in the document]
Comment (wendigo) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplit.java]: Much better. Thanks!

**Review by wendigo — DISMISSED** [review]
Please revert NodeManager changes

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java [hunk 8]
```diff
@@ -81,11 +80,10 @@ public BigQuerySplitManager(
     {
         this.bigQueryClientFactory = requireNonNull(bigQueryClientFactory, "bigQueryClientFactory cannot be null");
         this.bigQueryReadClientFactory = requireNonNull(bigQueryReadClientFactory, "bigQueryReadClientFactory cannot be null");
-        this.parallelism = config.getParallelism();
+        this.actualParallelism = config.getParallelism().orElseGet(() -> requireNonNull(nodeManager, "nodeManager cannot be null").getRequiredWorkerNodes().size());
```
Comment (wendigo) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:90]: If we want to do it this way we should use memoized supplier

Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java]: Let's just revert the commit.

**Review by wendigo — APPROVED** [review]

**Pull request merged**